src/main/java/com/spotify/folsom/client/Request.java

```diff
@@ -15,30 +15,37 @@
  */
 package com.spotify.folsom.client;
 
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Lists;
 import com.google.common.util.concurrent.AbstractFuture;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufAllocator;
 
 import java.io.IOException;
 import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.util.List;
 
 import static com.google.common.base.Preconditions.checkNotNull;
 
 public abstract class Request<V> extends AbstractFuture<V> {
-  protected final String key;
+  protected final byte[] key;
   protected final int opaque;
 
-  public Request(String key, int opaque) {
-    this.key = checkNotNull(key, "key");
+  protected Request(String key, Charset charset, int opaque) {
+    this(encodeKey(key, charset), opaque);
+  }
+
+  protected Request(byte[] key, int opaque) {
+    this.key = key;
     this.opaque = (opaque << 8) & 0xFFFFFF00;
-    Utils.validateKey(key);
   }
 
   protected static ByteBuf toBuffer(final ByteBufAllocator alloc, ByteBuffer dst) {
     return toBuffer(alloc, dst, 0);
   }
 
-  public String getKey() {
+  public byte[] getKey() {
```

Comment:
True, though this class is not really directly exposed to regular users of the library.

We could of course try to improve it by making it package private and move all the classes that need it into the same package. Which might make sense anyway.